### PR TITLE
Allow user defined exceptions to have docstrings

### PR DESCRIPTION
**Pull request by mejrs**
For #2016

### CODE REVIEW

File: src/err/mod.rs
Comment (mejrs) [inline src/err/mod.rs:360]: We could also just merge this and new_doc, if a breaking change is acceptable here.

File: src/exceptions.rs
Comment (mejrs) [inline src/exceptions.rs:202]: We could also have a default docstring. I chose not to because I'm not sure whether we should write things that may be visible to end users.

Comment (davidhewitt) [reply src/exceptions.rs:202]: Yeah I agree that we should not generate a default docstring.

File: src/err/mod.rs
Comment (davidhewitt) [inline src/err/mod.rs:345]: Mmm good idea checking for errors. If we're breaking `new_type` anyway, maybe we should change it to return `PyResult<Py<PyType>>` - this pushes the error handling downstream and also makes the pointer ownership clearer?

File: src/exceptions.rs
Comment (davidhewitt) [inline src/exceptions.rs:173]: You might find it works if set locals instead of globals?

File: src/err/mod.rs
Comment (davidhewitt) [reply src/err/mod.rs:360]: Hmmm. I'm really on the fence about this. I do think that a `_with_doc` method should probably take `&str` rather than `Option<&str>`. These two methods can probably share a common implementation which takes option.

Otoh, I feel more strongly that the return value of `new_type` should be changed (see other comment). So that does open up the chance for us to make more breakages.

Which design do you prefer? Single method taking option or two methods?

**Comment by davidhewitt** [comment]
Sorry it took me a while to get to this, thanks for implementing! Got a few ideas and comments r.e. the `new_type` stuff

File: src/err/mod.rs
Comment (mejrs) [reply src/err/mod.rs:360]: I'd definitely prefer a single method that takes an option.

File: src/exceptions.rs
Comment (mejrs) [reply src/exceptions.rs:173]: I think so too, I just find it weird it works intermittently like this.

**Comment by mejrs** [comment]
I have combined the functions and made it return Result<Py<PyAny>>. I think this is good to go after the 1.57 changes merge.

**Comment by davidhewitt** [comment]
Sorry for the catastrophically slow reviewing. Looks great to me! Just needs rebasing and a CHANGELOG entry :)

**Comment by mejrs** [comment]
> Sorry for the catastrophically slow reviewing. Looks great to me! Just needs rebasing and a CHANGELOG entry :)

Should be good to go now :)

**Review by davidhewitt — APPROVED** [review]
Perfect, thanks again!

**Comment by 9999years** [comment]
Any chance we could cut a release with this feature? I forked pyo3 to add this but it's already here, just unreleased.

**Comment by davidhewitt** [comment]
@9999years please follow #2079 - we're getting close to last piece we want to complete before next release lands. You might be able to use a git dependency in the meanwhile?

**Pull request merged**
